File: multiple-languages/dotnet/AlibabaCloud.SDK.ROS.CDK.Alb/AlibabaCloud/SDK/ROS/CDK/Alb/ServerGroupProps.cs
using Amazon.JSII.Runtime.Deputy;

#pragma warning disable CS0672,CS0809,CS1591

namespace AlibabaCloud.SDK.ROS.CDK.Alb
{
    #pragma warning disable CS8618

    /// <summary>Properties for defining a `ALIYUN::ALB::ServerGroup`.</summary>
    [JsiiByValue(fqn: "@alicloud/ros-cdk-alb.ServerGroupProps")]
    public class ServerGroupProps : AlibabaCloud.SDK.ROS.CDK.Alb.IServerGroupProps
    {
        /// <summary>Property healthCheckConfig: The configuration of health checks.</summary>
        [JsiiProperty(name: "healthCheckConfig", typeJson: "{\"union\":{\"types\":[{\"fqn\":\"@alicloud/ros-cdk-core.IResolvable\"},{\"fqn\":\"@alicloud/ros-cdk-alb.RosServerGroup.HealthCheckConfigProperty\"}]}}")]
        public object HealthCheckConfig
        {
            get;
            set;
        }

        /// <summary>Property serverGroupName: The name of the server group.</summary>
        /// <remarks>
        /// The name must be 2 to 128 characters in length, and
        /// can contain letters, digits, periods (.), underscores (_), and hyphens (-). The name
        /// must start with a letter.
        /// </remarks>
        [JsiiProperty(name: "serverGroupName", typeJson: "{\"union\":{\"types\":[{\"primitive\":\"string\"},{\"fqn\":\"@alicloud/ros-cdk-core.IResolvable\"}]}}")]
        public object ServerGroupName
        {
            get;
            set;
        }

        /// <summary>Property vpcId: The ID of the virtual private cloud (VPC).</summary>
        /// <remarks>
        /// You can add only servers that are deployed
        /// in the specified VPC to the server group.
        /// Note This parameter is required if the ServerGroupType parameter is set to Instance or Ip.
        /// </remarks>
        [JsiiProperty(name: "vpcId", typeJson: "{\"union\":{\"types\":[{\"primitive\":\"string\"},{\"fqn\":\"@alicloud/ros-cdk-core.IResolvable\"}]}}")]
        public object VpcId
        {
            get;
            set;
        }

        /// <summary>Property protocol: The server protocol.</summary>
        /// <remarks>
        /// Valid values:
        /// HTTP: allows you to associate HTTPS, HTTP, or QUIC listeners with backend servers. This
        /// is the default value.
        /// HTTPS: allows you to associate HTTPS listeners with backend servers.
        /// </remarks>
        [JsiiOptional]
        [JsiiProperty(name: "protocol", typeJson: "{\"union\":{\"types\":[{\"primitive\":\"string\"},{\"fqn\":\"@alicloud/ros-cdk-core.IResolvable\"}]}}", isOptional: true)]
        public object? Protocol
        {
            get;
            set;
        }

        /// <summary>Property resourceGroupId: The ID of the resource group.</summary>
        [JsiiOptional]
        [JsiiProperty(name: "resourceGroupId", typeJson: "{\"union\":{\"types\":[{\"primitive\":\"string\"},{\"fqn\":\"@alicloud/ros-cdk-core.IResolvable\"}]}}", isOptional: true)]
        public object? ResourceGroupId
        {
            get;
            set;
        }

        /// <summary>Property scheduler: The scheduling algorithm.</summary>
        /// <remarks>
        /// Valid values:
        /// Wrr: Backend servers that have higher weights receive more requests than those that have
        /// lower weights. This is the default value.
        /// Wlc: Requests are distributed based on the weight and load of each backend server. The
        /// load refers to the number of connections to a backend server. If multiple backend
        /// servers have the same weight, requests are routed to the backend server with the least
        /// connections.
        /// Sch: specifies consistent hashing that is based on source IP addresses.
        /// </remarks>
        [JsiiOptional]
        [JsiiProperty(name: "scheduler", typeJson: "{\"union\":{\"types\":[{\"primitive\":\"string\"},{\"fqn\":\"@alicloud/ros-cdk-core.IResolvable\"}]}}", isOptional: true)]
        public object? Scheduler
        {
            get;
            set;
        }

        /// <summary>Property serverGroupType: The type of the server group.</summary>
        /// <remarks>
        /// Valid values:
        /// Instance: a server group that consists of servers. You can add Elastic Compute Service (ECS)
        /// instances, elastic network interfaces (ENIs), and elastic container instances to this
        /// type of server group. This is the default value.
        /// Ip: a server group that consists of IP addresses. You can add IP addresses to this type
        /// of server group.
        /// </remarks>
        [JsiiOptional]
        [JsiiProperty(name: "serverGroupType", typeJson: "{\"union\":{\"types\":[{\"primitive\":\"string\"},{\"fqn\":\"@alicloud/ros-cdk-core.IResolvable\"}]}}", isOptional: true)]
        public object? ServerGroupType
        {
            get;
            set;
        }

        /// <summary>Property stickySessionConfig: The configuration of session persistence.</summary>
        /// <remarks>
        /// Note This parameter is required if the ServerGroupType parameter is set to Instance or Ip.
        /// </remarks>
        [JsiiOptional]
        [JsiiProperty(name: "stickySessionConfig", typeJson: "{\"union\":{\"types\":[{\"fqn\":\"@alicloud/ros-cdk-core.IResolvable\"},{\"fqn\":\"@alicloud/ros-cdk-alb.RosServerGroup.StickySessionConfigProperty\"}]}}", isOptional: true)]
        public object? StickySessionConfig
        {
            get;
            set;
        }
    }
}

File: multiple-languages/go/alicloudroscdkalb/ServerGroupProps.go
package alicloudroscdkalb


// Properties for defining a `ServerGroup`.
//
// See https://www.alibabacloud.com/help/ros/developer-reference/aliyun-alb-servergroup
type ServerGroupProps struct {
	// Property healthCheckConfig: The configuration of health checks.
	HealthCheckConfig interface{} `field:"required" json:"healthCheckConfig" yaml:"healthCheckConfig"`
	// Property serverGroupName: The name of the server group.
	//
	// The name must be 2 to 128 characters in length, and
	// can contain letters, digits, periods (.), underscores (_), and hyphens (-). The name
	// must start with a letter.
	ServerGroupName interface{} `field:"required" json:"serverGroupName" yaml:"serverGroupName"`
	// Property vpcId: The ID of the virtual private cloud (VPC).
	//
	// You can add only servers that are deployed
	// in the specified VPC to the server group.
	// Note: This parameter is required if the ServerGroupType parameter is set to Instance or Ip.
	// Note: This parameter takes effect when the ServerGroupType parameter is set to Instance or Ip.
	VpcId interface{} `field:"required" json:"vpcId" yaml:"vpcId"`
	// Property connectionDrainConfig: Configuration related to graceful connection interruption.Enable graceful connection interruption. After the backend server is removed or the health check fails, the load balancing allows the existing connection to be transmitted normally within a certain period of time.Note:  Basic Edition instances do not support enabling graceful connection interruption. Only Standard Edition and WAF Enhanced Edition instances support it.Server type and IP type server group support graceful connection interruption. Function Compute type does not support it.
	ConnectionDrainConfig interface{} `field:"optional" json:"connectionDrainConfig" yaml:"connectionDrainConfig"`
	// Property protocol: The backend protocol.
	//
	// Valid values:
	// HTTP (default): The server group can be associated with HTTPS, HTTP, and QUIC listeners.
	// HTTPS: The server group can be associated with HTTPS listeners.
	// gRPC: The server group can be associated with HTTPS and QUIC listeners.
	// Note: If the ServerGroupType parameter is set to Fc, you can set Protocol only to HTTP.
	Protocol interface{} `field:"optional" json:"protocol" yaml:"protocol"`
	// Property resourceGroupId: The ID of the resource group.
	ResourceGroupId interface{} `field:"optional" json:"resourceGroupId" yaml:"resourceGroupId"`
	// Property scheduler: The scheduling algorithm.
	//
	// Valid values:
	// Wrr (default): The weighted round-robin algorithm is used. Backend servers that have higher weights receive more requests than those that have lower weights.
	// Wlc: The weighted least connections algorithm is used. Requests are distributed based on the weights and the number of connections to backend servers. If two backend servers have the same weight, the backend server that has fewer connections is expected to receive more requests.
	// Sch: The consistent hashing algorithm is used. Requests from the same source IP address are distributed to the same backend server.
	// Note: This parameter takes effect when the ServerGroupType parameter is set to Instance or Ip.
	Scheduler interface{} `field:"optional" json:"scheduler" yaml:"scheduler"`
	// Property serverGroupType: The type of the server group.
	//
	// Valid values:
	// Instance (default): allows you add servers by specifying Ecs, Ens, or Eci.
	// Ip: allows you to add servers by specifying IP addresses.
	// Fc: allows you to add servers by specifying functions of Function Compute.
	ServerGroupType interface{} `field:"optional" json:"serverGroupType" yaml:"serverGroupType"`
	// Property serviceName: This parameter is available only if the ALB Ingress controller is used.
	//
	// In this case, set the parameter to the name of the Kubernetes Service that is associated with the server group.
	ServiceName interface{} `field:"optional" json:"serviceName" yaml:"serviceName"`
	// Property slowStartConfig: Slow start related configuration.After slow start is enabled, the backend server newly added to the backend server group will be preheated within the set time period, and the number of requests forwarded to the server will increase linearly.Note: Basic Edition instances do not support slow start, only Standard Edition and WAF Enhanced Edition instances support it.Server type and IP type server groups support slow start configuration, but Function Compute type does not.Slow start can only be enabled when the backend scheduling algorithm is the weighted polling algorithm.
	SlowStartConfig interface{} `field:"optional" json:"slowStartConfig" yaml:"slowStartConfig"`
	// Property stickySessionConfig: The configuration of session persistence.
	//
	// Note: This parameter is required if the ServerGroupType parameter is set to Instance or Ip.
	StickySessionConfig interface{} `field:"optional" json:"stickySessionConfig" yaml:"stickySessionConfig"`
	// Property tags: Tags to attach to instance.
	//
	// Max support 20 tags to add during create instance. Each tag with two properties Key and Value, and Key is required.
	Tags *[]*RosServerGroup_TagsProperty `field:"optional" json:"tags" yaml:"tags"`
	// Property uchConfig: URL consistency hash parameter configuration.
	UchConfig interface{} `field:"optional" json:"uchConfig" yaml:"uchConfig"`
	// Property upstreamKeepaliveEnabled: Whether to enable upstream keepalive.
	UpstreamKeepaliveEnabled interface{} `field:"optional" json:"upstreamKeepaliveEnabled" yaml:"upstreamKeepaliveEnabled"`
}

